### ISSUE TITLE
Add permissions for viewing blocked user account profiles (and/or their content)

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#6123]
## Description of the need

It would be helpful to have additional permissions that control whether a role can view

(a) the user profile of a blocked user
(b) any content that had been created by a blocked user

The need arose in connection with backdropcms.org, where potential spam accounts are (manually) blocked (based on various indicators of potential spamminess), and then blocked accounts are automatically deleted after two weeks. (The reason for not deleting them immediately is to avoid deleting a valid account that just looked spammy; with the delayed deletion, it gives the owners of real accounts a chance to flag the improper blockage before everything goes away.)

During the 2-week blockage, the user accounts are still accessible by anonymous users and show up in Google searches, etc. It would be nice if during the "penalty box" of account blockage, the blocked account was not publicly accessible. And along the same lines, if a blocked account had created additional content, it would be nice to have the option of blocking content that had been created by blocked accounts (similarly to the existing ability to delete content from an account when that account is deleted).

## Proposed solution

In the **User** permissions section, there is a "View user profiles" permission. We could add a permission "View blocked user profiles" that is initially checked by default (to match current behavior) but that could be unchecked to hide blocked user profiles. We'd then need to change "View user profiles" to "View unblocked user profiles" (at least the display name, but presumably not the machine name).

In the **Node** permisisons section, there is a "View published content" permssion. We could add a permission "View published content from blocked users" that (like the above) is initially checked by default (to match current behavior). Same considerations would apply, e.g., making the original permission "View published content from unblocked users".

There may well be better names and splitting of permission functionality than the above; suggestions solicited.

## Alternatives that have been considered

In the case of backdropcms.org, an alternative would be just a change of policy: if a user account is so obviously spammy that we don't want it visible, we could just go ahead and delete it immediately, and only use the "block-now, delete-later" approach for questionable accounts. But it seems like having the ability to distinguish visibility between blocked and active accounts would be a good and useful thing overall.

H/T @olafgrabienski.

**Comment by kiamlaluno** [comment]
I thought that blocked accounts would only be visible to user administrators, but then I noticed that [`User::access()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21user%21user.entity.inc/function/User%3A%3Aaccess/1) is using the following code.

```php
if ($op == 'view') {
  // Admins can view all, users can view own profiles at all times.
  if ($account->uid == $this->uid || user_access('administer users', $account)) {
    $rights[$account->uid][$cid][$op] = TRUE;
    return $rights[$account->uid][$cid][$op];
  }
  elseif (user_access('access user profiles', $account)) {
    $rights[$account->uid][$cid][$op] = $account->status;
    return $rights[$account->uid][$cid][$op];
  }
  $rights[$account->uid][$cid][$op] = FALSE;
  return $rights[$account->uid][$cid][$op];
}
```

The code does not check the value of `$this->status`. When anonymous users (like search crawlers) have the *access user profiles* permission, they can see blocked accounts too.
I am not sure this is the expected result.

Comparing Drupal 7 [`user_view_access()`](https://api.drupal.org/api/drupal/modules%21user%21user.module/function/user_view_access/7.x) with Backdrop [`user_view_access()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21user%21user.module/function/user_view_access/1) I see the latter is missing the check for the account status.

```php
/**
 * Drupal 7
 */
function user_view_access($account) {
  $uid = is_object($account) ? $account->uid : (int) $account;

  // Never allow access to view the anonymous user account.
  if ($uid) {

    // Admins can view all, users can view own profiles at all times.
    if ($GLOBALS['user']->uid == $uid || user_access('administer users')) {
      return TRUE;
    }
    elseif (user_access('access user profiles')) {

      // At this point, load the complete account object.
      if (!is_object($account)) {
        $account = user_load($uid);
      }
      return is_object($account) && $account->status;
    }
  }
  return FALSE;
}
```

```php
/**
 * Backdrop
 */
function user_view_access($account) {
  if ($account instanceof User) {
    return $account && $account->access('view');
  }
  return FALSE;
}
```

I am favorable to add a permission to see the blocked accounts, even if changing the code to check if the viewed account is blocked would mitigate what reported here. (Blocked accounts will not be accessible to search crawlers.)

**Comment by olafgrabienski** [comment]
Thanks for opening this issue, @bugfolder!

> In the User permissions section, there is a "View user profiles" permission. We could add a permission "View blocked user profiles" that is initially checked by default (to match current behavior) [...]

IIRC, on a vanilla Backdrop site the "View user profiles" permission is only enabled for the Admin role, so there is no need to enable the new "View blocked user profiles" permission (for the anonymous role).

**Comment by kiamlaluno** [comment]
> IIRC, on a vanilla Backdrop site the "View user profiles" permission is only enabled for the Admin role

Yes, but most sites want to make user profiles visible to everybody with an account on the sites. I cannot say how much sites also allow access to view user profiles also to anonymous users, just to make the user profile pages accessible to search engines.

**Comment by klonos** [comment]
I also support this change 👍🏼 

> IIRC, on a vanilla Backdrop site the "View user profiles" permission is only enabled for the Admin role ...

> Yes, but most sites want to make user profiles visible to everybody with an account on the site. ...

Sounds like we may need an update hook then. On existing sites, we'll need to iterate through all user roles, and then check if the permission to view profiles is set for the role. If so, then we should also enable the new permissions (in order to retain the current functionality).

Having said the above, I would still argue that current sites may be in the same situation as our b.org site: they have enabled the existing permission to have all user profiles visible for SEO purposes and the nature of the site (community etc.), but the current lack of a specific permission to block access to blocked user accounts makes it a dead end - had the site owners/admins had such a permission available to them, I believe that the sensible option would be to only allow site admins/moderators to be able to view blocked account profiles 🤔 ...perhaps security-wise we need to make a conscious choice to "break" existing sites, and then add a change record and a notice in the release notes that people should review their current permissions manually and make a choice.

This last bit would actually make it. good candidate for #3624 😉

**Comment by kiamlaluno** [comment]
To make my previous comment clearer: Most of the sites would probably want to make user profiles visible to people with an account on the site, except in the case an account is blocked.
